File: heap/lb-10-merge-k-sorted-list.py
# Given k sorted lists
# merge using heap min heap or priority queue
# take first k elements of all the lists and build heap.
# extract min, and add next smallest element and advance pointer.
# do the same thing for complete lists
# O(NlogK) = >N of elements, K= number of lists, Build heap O(K)
import heapq
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def __lt__(self, other):
        return self.val < other.val

def merge_k_sorted_lists(lists):
    head = temp = ListNode(0)
    heap = []
    for lst in lists:
        if lst:
            heapq.heappush(heap, (lst.val, lst))
    while len(heap):
        val, node = heapq.heappop(heap)
        temp.next = ListNode(val)
        temp = temp.next
        node = node.next
        if node:
            heapq.heappush(heap, (node.val, node))
    return head

                



l1 = ListNode(1)
l1.next = ListNode(4)
l1.next.next = ListNode(5)

l2 = ListNode(1)
l2.next = ListNode(3)
l2.next.next = ListNode(4)

l3 = ListNode(2)
l3.next = ListNode(6)
lists = [l1, l2, l3]

merged = merge_k_sorted_lists(lists)
# setattr(Node, "__lt__", lambda self, other: self.data < other.data)
# if node class is not accessible
while merged:
    print(f"{merged.val}", end="->")
    merged = merged.next
print("NULL")

File: heap/lb-10-merge-k-sorted-list.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;
class Node {
   public:
    int data;
    Node *next;
    Node(int val) {
        data = val;
        next = NULL;
    }
};
// custom comparator
class Comp {
   public:
    bool operator()(pair<int, Node *> p1, pair<int, Node *> p2) {
        return p1.first > p2.first;
    }
};

Node *merge(vector<Node *> lists) {
    Node *head = new Node(0);
    Node *temp = head;
    priority_queue<pair<int, Node *>, vector<pair<int, Node *>>, Comp> heap;  // min heap
    // push each list head in heap
    for (int i = 0; i < lists.size(); i++) {
        // atleast one node is present in each list : assumed
        heap.push({lists[i]->data, lists[i]});
    }
    // now each node as per value and merge to list
    while (!heap.empty()) {
        pair<int, Node *> top = heap.top();
        heap.pop();
        temp->next = new Node(top.first);
        temp = temp->next;
        Node *node = top.second->next;
        if (node) {
            heap.push({node->data, node});
        }
    }
    // node point to dummy node
    // so original list is
    return head->next;
}

int main() {
    // 1-2-3
    // 3-5-7
    // 1-2-3-3-5-7
    Node *l1 = new Node(1);
    l1->next = new Node(2);
    l1->next->next = new Node(300);
    Node *l2 = new Node(3);
    l2->next = new Node(5);
    l2->next->next = new Node(7);
    vector<Node *> lists = {l1, l2};
    Node *mr = merge(lists);
    Node *temp = mr;
    while (temp) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
    return 0;
}
